core/src/apps/base.py

```diff
@@ -183,6 +183,64 @@ async def handle_DoPreauthorized(
     return await handler(ctx, req, authorization.get())  # type: ignore [Expected 2 positional arguments]
 
 
+async def handle_UnlockPath(ctx: wire.Context, msg: UnlockPath) -> protobuf.MessageType:
+    from trezor.crypto import hmac
+    from trezor.messages import Cancel, UnlockedPathRequest
+    from trezor.ui.layouts import confirm_action
+    from apps.common.paths import SLIP25_PURPOSE
+    from apps.common.seed import Slip21Node, get_seed
+    from apps.common.writers import write_uint32_le
+
+    _KEYCHAIN_MAC_KEY_PATH = [b"TREZOR", b"Keychain MAC key"]
+
+    # UnlockPath is relevant only for SLIP-25 paths.
+    # Note: Currently we only allow unlocking the entire SLIP-25 purpose subtree instead of
+    # per-coin or per-account unlocking in order to avoid UI complexity.
+    if len(msg.address_n) != 1 or msg.address_n[0] != SLIP25_PURPOSE:
+        raise wire.DataError("Invalid path")
+
+    seed = await get_seed(ctx)
+    node = Slip21Node(seed)
+    node.derive_path(_KEYCHAIN_MAC_KEY_PATH)
+    mac = utils.HashWriter(hmac(hmac.SHA256, node.key()))
+    for i in msg.address_n:
+        write_uint32_le(mac, i)
+    expected_mac = mac.get_digest()
+
+    # Require confirmation to access SLIP25 paths unless already authorized.
+    if msg.mac:
+        if len(msg.mac) != len(expected_mac) or not utils.consteq(
```

Comment:
the length check is already covered by `consteq`, no?

Reply:
No, the caller needs to guarantee that `len(sec) >= len(pub)`. This is important in case `len(sec)` is itself secret information. I actually think this is the most dangerous function that we have in python.

https://github.com/trezor/trezor-firmware/blob/0b4ccf45fc5a7de9ccff48dd33c8458bff1d20c9/core/embed/extmod/modtrezorutils/modtrezorutils.c#L50-L75

We could make it better by:
1. Adding `assert(secbuf.len >= pubbuf.len);` at the beginning, so that we at least have a better chance of catching programming bugs in the debug build.
2. Iterating only up to the minimum length, which can be computed in constant time:
    ```c
    // Compute minlen = min(secbuf.len, pubbuf.len) in constant time.
    uint32_t mask = -(secbuf.len < pubbuf.len);
    uint32_t minlen = (secbuf.len & mask) | (pubbuf.len & ~mask);
    ```
    However, if `secbuf.len < pubbuf.len`, then the timing of the loop reveals the length of `secbuf`. This is a problem if the caller actually knows that they can read past the buffer's length, because more bytes are allocated, e.g. in case of `bytearray`. BTW, in that case we could check `mp_obj_array_t::free`, but not with `mp_buffer_info_t` directly. The question is whether there actually is any place in our code that uses this feature.
3. Accessing `s` modulo `secbuf.len`, i.e.
    ```c
    diff |= s[i % secbuf.len] - p[i];
    ```
    Not sure how to deal with `secbuf.len==0` though. Also not sure whether the `%` operation is constant-time.

Reply:
I created a separate issue for this https://github.com/trezor/trezor-firmware/issues/2466/.